### PR TITLE
fix(consul): preserve membership consistency

### PR DESCRIPTION
**Pull request by ReubenBond**
## Problem
Consul membership reads can return stale views, cleanup can remove non-Dead or concurrently updated registrations, and transaction operation failures can be mistaken for ordinary contention.

## Solution
Use consistent canonical reads and atomic row/table CAS transactions for membership inserts and updates. Keep single-silo reads scoped to the registration/heartbeat prefix, bracketed by consistent version-key reads. Return the current table version for missing rows, propagate backend failures, and isolate cluster prefixes. Distinguish native stale-index conflicts from other transaction errors, including ACL/storage failures returned inside HTTP 409 responses.

Each silo owns periodic `IAmAlive` publication during its registered lifetime. Each periodic update performs one blind native PUT to the existing heartbeat-key layout, preserving registration fields, the row ETag, and the table version. Canonical insert/update transactions access only the registration and version keys, so an owner's timestamp publication leaves previously read membership tokens usable. A row or version conflict returns false to the caller; version-read retries use cancellation-aware backoff.

Prune only Dead registrations whose latest start, heartbeat, and vote precede the cutoff, using paired row/heartbeat DeleteCAS while retaining the table version and supporting legacy rows without heartbeat keys. Bulk reads and cleanup index each returned snapshot once for heartbeat association. Initialization uses a single native create-if-absent CAS.

Enable the existing strong-version assertions and add focused native Consul SDK regressions with a bounded HTTP fake, including exact request counts, field/version isolation, canonical token usability across heartbeat publication, and operation-error propagation. Keep the existing storage layout and configuration surface, with a small Consul documentation clarification.

Includes only the shared legacy-cleanup preparation commit from #11308 (`bede893d290a9035f196f4c3ed58213d102efc6c`, cherry-picked as `a4b8679173460306e8cc626d2709db9e5037d59c`) before the focused Consul commits. The conformance kit and adapters remain in the later PR.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/orleans/pull/11314)

### CODE REVIEW

File: src/Orleans.Clustering.Consul/ConsulBasedMembershipTable.cs [hunk 1]
```diff
@@ -175,23 +178,36 @@ public async Task<bool> InsertRowAsync(MembershipEntry entry, TableVersion table
...
+                        new(updateKV.Key, KVTxnVerb.CAS) { Index = expectedIndex, Value = updateKV.Value },
+                        GetVersionRowUpdate(tableVersion),
+                        new(heartbeat.Key, KVTxnVerb.CAS) { Index = currentHeartbeat?.ModifyIndex ?? 0, Value = heartbeat.Value }
+                    };
 
-                return true;
+                    if ((await _consulClient.KV.Txn(operations, cancellationToken)).Response.Success)
+                    {
+                        return true;
+                    }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Clustering.Consul/ConsulBasedMembershipTable.cs]: Fixed in fe713fc2163a43bf83cab351366a74f510d41275. Both UpdateRowAsync and UpdateIAmAliveAsync now await a cancellation-aware 100 ms delay after a transaction conflict. Successful writes still return immediately. Added repeated-conflict and cancellation coverage for both paths; the net10 build and all 40 focused Consul tests pass.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

File: src/Orleans.Clustering.Consul/ConsulBasedMembershipTable.cs [hunk 2]
```diff
@@ -247,25 +290,25 @@ private static TableVersion GetTableVersion(string? versionKey, QueryResult<KVPa
...
 
-        private async Task<(ConsulSiloRegistration?, TableVersion)> GetConsulSiloRegistration(SiloAddress siloAddress, CancellationToken cancellationToken)
+        private async Task<(ConsulSiloRegistration?, TableVersion, KVPair?)> GetConsulSiloRegistration(SiloAddress siloAddress, CancellationToken cancellationToken)
         {
             var deploymentKey = ConsulSiloRegistrationAssembler.FormatDeploymentKVPrefix(this.clusterId, this.kvRootFolder);
             var siloKey = ConsulSiloRegistrationAssembler.FormatDeploymentSiloKey(this.clusterId, this.kvRootFolder, siloAddress);
-            var entries = await _consulClient.KV.List(deploymentKey, cancellationToken);
-            if (entries.Response == null) return (null, NotFoundTableVersion);
+            var entries = await _consulClient.KV.List(deploymentKey + "/", ConsistentRead, cancellationToken);
+            if (entries.Response == null) return (null, NotFoundTableVersion, null);
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Clustering.Consul/ConsulBasedMembershipTable.cs]: Fixed in fe713fc2163a43bf83cab351366a74f510d41275. GetConsulSiloRegistration now reads the exact version key, lists only the requested silo's registration/heartbeat prefix, then re-reads the version key. All three reads use Consistent mode; matching version ModifyIndex values place the atomic row/heartbeat snapshot within that table version. A changed ETag retries after a cancellation-aware 100 ms delay. This preserves missing-row and unversioned Dead-pruning behavior while removing the full-cluster payload. The SDK's KV.Txn helper exposes only WriteOptions, so it cannot request consistent read-only transactions directly. Added regressions for bounded read scope with 50 silos, version changes before/after the row snapshot, and final version-read errors. The net10 build and all 40 focused tests pass.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

**Comment by ReubenBond** [comment]
Addressing both body-only findings in [this review](https://github.com/dotnet/orleans/pull/11314#pullrequestreview-5248826500): **the premise is incorrect; the existing DeleteCAS implementation and fake match Consul's native semantics.**

- In the CI server's Consul 1.19 line, [`kvsDeleteCASTxn`](https://github.com/hashicorp/consul/blob/bf0166d85082f384a94c5c0e6227619e63f3c644/agent/consul/state/kvs.go#L318-L342) returns `entry == nil, nil` when the entry is absent or its index differs. Absence therefore returns success; an existing key with a mismatching index returns false. The same code is present in [1.19.2](https://github.com/hashicorp/consul/blob/048f1936f5ab43bb7b5096dfdca81db05a19b9cf/agent/consul/state/kvs.go#L318-L342).
- The [transaction dispatcher](https://github.com/hashicorp/consul/blob/bf0166d85082f384a94c5c0e6227619e63f3c644/agent/consul/state/txn.go#L26-L31) calls that function and generates a conflict only when it returns false. This is the transaction path used here, not merely the standalone KV endpoint.
- Consul's own [`TestStateStore_KVSDeleteCAS`](https://github.com/hashicorp/consul/blob/bf0166d85082f384a94c5c0e6227619e63f3c644/agent/consul/state/kvs_test.go#L634-L642) explicitly requires a nonexistent-key CAS delete to return `(true, nil)`.
- Our pinned [Consul SDK 1.7.14.10](https://github.com/G-Research/consuldotnet/blob/v1.7.14.10/Consul/KV.cs#L610-L626) forwards the transaction to `/v1/txn` and reports HTTP 200 as success.

Consequently, heartbeat `DeleteCAS(Index=0)` succeeds while the key remains absent and conflicts if a concurrent writer creates it with a nonzero ModifyIndex. The paired transaction then rolls back the row deletion. The fake's `current is null || current.ModifyIndex == operation.Index` models this rule correctly. Requiring key existence in the fake would introduce the mismatch alleged by the review. No implementation change is needed.

Existing focused regressions cover legacy missing-heartbeat pruning, concurrent heartbeat creation preserving the row, retired-heartbeat races, and backend-error propagation. Re-ran the focused SDK suite: **40 passed, zero failures/skips**. On unchanged head `fe713fc2163a43bf83cab351366a74f510d41275`, the Consul backend CI jobs also passed **54 tests each**, zero failures/skips: [net8.0](https://github.com/dotnet/orleans/actions/runs/35355050650/job/105632252965), [net10.0](https://github.com/dotnet/orleans/actions/runs/35355050650/job/105632253262). The exact missing-key semantics are established by the server implementation and upstream assertion above; the local race regressions exercise our fake against those semantics.

File: test/Extensions/Orleans.Clustering.Consul.Tests/ConsulMembershipTableConsistencyTests.cs [hunk 3]
```diff
@@ -0,0 +1,785 @@
+using System.Globalization;
+using System.Net;
+using System.Text;
+using Microsoft.Extensions.Logging.Abstractions;
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Extensions/Orleans.Clustering.Consul.Tests/ConsulMembershipTableConsistencyTests.cs:4]: This file is declared in `namespace Consul.Tests;`, so C# resolves these SDK types from the enclosing `Consul` namespace without an additional `using`. The current head (`fe713fc2163a43bf83cab351366a74f510d41275`) compiled successfully, and both native Consul CI jobs passed on .NET 8 and .NET 10. No code change is needed here.

<sub>Auto-replied by the [GitHub Copilot app](https://gh.io/app)</sub>

**Comment by ReubenBond** [comment]
Updated in dbcea6b37c59dbe6d887835a20f4c4e237e2b305 for the clarified ownership contract: the owning silo publishes periodic IAmAlive updates during its registered lifetime using exactly one native KV.Put to its heartbeat key. Cancellation and storage failures propagate, including an explicit failure for a false write response.

The earlier periodic multi-writer/max-timestamp and post-retirement no-op expectations have been replaced with exact one-PUT/zero-read/zero-transaction, timestamp-encoding, and field/version-isolation regressions. Full membership-row transactions, their conflict backoff, scoped coherent reads, and cleanup CAS remain protected. This supersedes the previous extra periodic-heartbeat protections while retaining the applicable full-row review fixes; the resolved DeleteCAS and namespace findings remain unchanged.

The net10 build and all 41 focused Consul tests pass. Documentation and PR description now describe the owner-write contract. CI is being checked on this exact new head.

**Comment by ReubenBond** [comment]
Follow-up cb80e7469429537e9cde7ff32b6e5c213a2a5cb9 applies the final canonical-token clarification: insert and update transactions now contain only registration CAS and table-version CAS. They perform zero prerequisite reads and never access, compare, or write the heartbeat key. An owner's one-PUT heartbeat update therefore leaves both previously read membership tokens usable. Full-row conflicts return false directly to the caller instead of retrying internally; the consistent ReadRow version fence still retries canonical version changes with backoff.

The revised native SDK tests cover status and vote updates using original tokens after an owner PUT and with an owner PUT interleaved immediately before transaction application: exactly one canonical transaction, zero canonical read calls, zero heartbeat-triggered conflicts. They also verify that heartbeat changes at every ReadRow phase cause zero fence retries, and that insertion is independent of a pre-existing heartbeat key. Cleanup's paired DeleteCAS remains scoped to Dead-row recency protection.

All 47 focused tests pass with the net10 build, and the documentation source audit passes. This supersedes the earlier full-row heartbeat-CAS/max behavior and the now-unnecessary UpdateRow conflict-retry loop; storage layout and native failure propagation are preserved. The PR description reflects the final contract. Checking CI on this exact new head; readiness and reviewer settings are unchanged.

**Comment by ReubenBond** [comment]
### Active cost/error reassessment

Reassessed against main `0f9537dde1c10107d059fc10f1171f6f04a43f19` (its Consul source is unchanged from this PR's original base). Commit `3b4b2c732a4dc4ba01f11142cf90f49b16549c84` fixes verified waste and an error-boundary issue. Here N is the number of registrations, M the returned KV count, and D the selected Dead candidates.

| Path in `ConsulBasedMembershipTable.cs` | Main | Before this reassessment | Current / rationale |
|---|---|---|---|
| `InitializeMembershipTableAsync` (L60) | No requests | 1 singleton transaction / 1 CAS op | 1 native CAS request, 0 transaction envelopes. Index zero performs atomic create-if-absent; existing versions remain intact. `false` initialization still costs zero. |
| `UpdateIAmAliveAsync` (L221) | 1 PUT | 1 PUT | Unchanged: one blind owner PUT, no reads/CAS/transactions/retries; one timestamp serialization. |
| `InsertRowAsync` / `UpdateRowAsync` (L149/L180) | 1 transaction / 2 CAS ops | Same | Unchanged: row + version only, no reads or provider retries. Row JSON encoded once; version encoded once; SDK serializes one transaction envelope. CAS zero is rejected on update because Consul treats it as insertion. |
| `ReadAllAsync` / `GetSiloEntries` (L123/L292) | 1 List; N*M heartbeat-key comparisons | Same request count, consistent mode | 1 List; one M-entry ordinal index and N lookups. Registration/heartbeat JSON decoding remains once per returned value. Gateway reads reuse this path. |
| `GetConsulSiloRegistration` (L308) | 1 full-cluster List | 3 consistent reads with silo-local payload | Retained: version Get / silo-prefix List / version Get. Compare **version-key ModifyIndex**, not heartbeat metadata. Retry only a canonical version change, with 100 ms cancellation-aware delay. |
| `CleanupDefunctSiloEntriesAsync` (L348) | 1 List + D DeleteTree; quadratic association | 1 List + D paired-DeleteCAS transactions | Same native counts, linear association. Each candidate has 2 ops, no version op, no point rereads/rescan/retry. A conflicting candidate leaves unrelated candidates eligible to progress. |
| `DeleteMembershipTableEntriesAsync` (L237) | 1 DeleteTree | 1 DeleteTree | Retained: exact requested cluster prefix including separator; a false response surfaces as failure. |

**Native guarantees:** Consul [KVSList](https://github.com/hashicorp/consul/blob/048f1936f5ab43bb7b5096dfdca81db05a19b9cf/agent/consul/state/kvs.go#L209-L225) uses a single read transaction; the KV API returns the prefix in one response, so ReadAll needs no pagination fence. [CAS](https://github.com/hashicorp/consul/blob/048f1936f5ab43bb7b5096dfdca81db05a19b9cf/agent/consul/state/kvs.go#L363-L387) checks the key's ModifyIndex (zero means absence), and [TxnRW](https://github.com/hashicorp/consul/blob/048f1936f5ab43bb7b5096dfdca81db05a19b9cf/agent/consul/state/txn.go#L430-L450) commits/rolls back all operations together. The bounded ReadRow fence remains necessary across its separate requests. The pinned SDK's KV.Txn only accepts WriteOptions; Raw.Write also uses WriteOptions, and its URL builder treats the endpoint as a path, not a way to pass `?consistent`. Avoided adding a custom transport to save those point reads. Consistent mode retains the required quorum-confirmed read boundary.

**Actual error bug fixed:** Consul [transaction prechecks](https://github.com/hashicorp/consul/blob/048f1936f5ab43bb7b5096dfdca81db05a19b9cf/agent/consul/txn_endpoint.go#L39-L57) put ACL denial in per-operation Errors, and the [HTTP handler](https://github.com/hashicorp/consul/blob/048f1936f5ab43bb7b5096dfdca81db05a19b9cf/agent/txn_endpoint.go#L345-L404) returns 409 for all transaction Errors. The SDK therefore exposes both contention and these failures with Success=false. `IsTransactionSuccessful` now accepts only the [native stale-index set/delete forms](https://github.com/hashicorp/consul/blob/048f1936f5ab43bb7b5096dfdca81db05a19b9cf/agent/consul/state/txn.go#L40-L56) as ordinary contention; other, mixed, and empty error responses throw with operation details. Unknown future error text also surfaces rather than being silently retried. Initialization separately surfaces the SDK's otherwise-absorbed unexpected 404.

**Retained deliberately:** existing JSON/key layout, missing-heartbeat StartTime behavior (also needed before a new owner's first publication), missing-version zero representation, strict invalid-version errors, and per-candidate cleanup atomicity. Wider cleanup transactions would couple rollback of independent candidates. No new locks, marker/schema sidecars, capability fallbacks, or rollout protocol. The separate heartbeat key naturally remains intact across canonical updates at zero extra native cost.

**Validation:** net10 compilation and 64 focused SDK tests passed, including native request shapes, 50-row heartbeat association, 10-candidate cleanup with one conflict and exactly one snapshot, singleton initialization, and 12 permission/storage/mixed/empty transaction-failure cases. Documentation audit and diff check passed. Native CI is being checked on this exact changed head. Request counts and algorithmic work were verified; no elapsed-time or production-throughput improvement is claimed.

**Comment by ReubenBond** [comment]
Exact-head validation on `3b4b2c732a4dc4ba01f11142cf90f49b16549c84`: the native Consul suites passed 78 tests each on net8.0 and net10.0, with zero failures/skips; the latest review reports no findings. The unrelated Windows/net8.0 BVT reminder identity-selection failure is tracked in #11325. The related merged ring/topology fixes are already ancestors of this head, and no newer matching fix was found. The first failed-jobs-only retry passed on unchanged code: .NET CI run 35373935320 attempt 2 is successful. The final checked PR rollup has no failed or pending checks. Readiness and reviewer settings remain unchanged.
